device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/twin/TwinMetadata.java

```diff
@@ -0,0 +1,269 @@
+     */
+    TwinMetadata(String lastUpdated, Integer lastUpdatedVersion, String lastUpdatedBy, String lastUpdatedByDigest) {
+        if (!Tools.isNullOrEmpty(lastUpdated)) {
+            /* SRS_TWIN_METADATA_21_001: [The constructor shall parse the provided `lastUpdated` String to the Date and store it as the TwinMetadata lastUpdated.] */
+            /* SRS_TWIN_METADATA_21_002: [The constructor shall throw IllegalArgumentException if it cannot convert the provided `lastUpdated` String to Date.] */
+            this.lastUpdated = ParserUtility.getDateTimeUtc(lastUpdated);
+        }
+
+        if (lastUpdatedBy != null) {
+            this.lastUpdatedBy = lastUpdatedBy;
+        }
+
+        if (lastUpdatedByDigest != null) {
+            this.lastUpdatedByDigest = lastUpdatedByDigest;
+        }
+
+        /* SRS_TWIN_METADATA_21_003: [The constructor shall store the provided lastUpdatedVersion as is.] */
+        this.lastUpdatedVersion = lastUpdatedVersion;
+
+        /* SRS_TWIN_METADATA_21_012: [The constructor shall throw IllegalArgumentException if both lastUpdated and lastUpdatedVersion are null.] */
+        if ((this.lastUpdatedVersion == null) && (this.lastUpdated == null)) {
+            throw new IllegalArgumentException("no valid data to create a TwinMetadata.");
+        }
+    }
+
+    /**
+     * CONSTRUCTOR (copy)
+     *
+     * <P> This private constructor will create a new instance of the TwinMetadata coping the information from the provided one.
+     *
+     * @param metadata the original {@code TwinMetadata} to copy.
+     */
+    TwinMetadata(TwinMetadata metadata) {
+        /* SRS_TWIN_METADATA_21_010: [The constructor shall throw IllegalArgumentException if the provided metadata is null.] */
+        if (metadata == null) {
+            throw new IllegalArgumentException("metadata to copy cannot be null");
+        }
+
+        /* SRS_TWIN_METADATA_21_011: [The constructor shall copy the content of the provided metadata.] */
+        this.lastUpdated = metadata.getLastUpdated();
+        this.lastUpdatedVersion = metadata.getLastUpdatedVersion();
+        this.lastUpdatedBy = metadata.getLastUpdatedBy();
+        this.lastUpdatedByDigest = metadata.getLastUpdatedByDigest();
+    }
+
+    /**
+     * Metadata extractor
+     *
+     * <p> This internal method will try to find $lastUpdated and $lastUpdatedVersion at the first
+     * level of the provided Map (Object), and create a new instance of the TwinMetadata with
+     * this information.
+     *
+     * <p> Once the provide Object can or cannot be a Map, and, if it is a Map, it can or cannot
+     * contains a valid metadata, this method contains the label <b>try</b>, which means that
+     * it can return a valid TwinMetadata or {@code null}.
+     *
+     * <p> For instance, for the follow Map, this method will create a TwinMetadata with
```

Comment:
```suggestion
     * <p> For instance, for the following Map, this method will create a TwinMetadata with
```

Reply:
I'll make this change in all the copies of this class, sure